File: packages/python-sdk/janua/http_client.py
"""HTTP client for making API requests with retry logic and error handling."""

import asyncio
import random
import time
from dataclasses import dataclass, field
from typing import Any, Callable, Dict, List, Optional, TypeVar, Union
import httpx
from httpx import Response, TimeoutException, NetworkError, AsyncClient

from .exceptions import (
    JanuaError,
    APIError,
    AuthenticationError,
    AuthorizationError,
    NotFoundError,
    ValidationError,
    RateLimitError,
    ServerError,
    NetworkConnectionError,
)
from .types import BaseResponse

T = TypeVar('T', bound=BaseResponse)


@dataclass
class RetryConfig:
    """Configuration for retry behavior.

    Attributes:
        max_attempts: Maximum number of retry attempts (including initial request)
        base_delay: Initial delay between retries in seconds
        max_delay: Maximum delay between retries in seconds
        exponential_base: Base for exponential backoff (default: 2.0)
        jitter: Whether to add random jitter to delay (prevents thundering herd)
        retry_on_status: HTTP status codes that should trigger a retry
        on_retry: Optional callback called before each retry
    """
    max_attempts: int = 3
    base_delay: float = 1.0
    max_delay: float = 60.0
    exponential_base: float = 2.0
    jitter: bool = True
    retry_on_status: List[int] = field(default_factory=lambda: [429, 500, 502, 503, 504])
    on_retry: Optional[Callable[[int, Exception, float], None]] = None

    def calculate_delay(self, attempt: int, error: Optional[Exception] = None) -> float:
        """Calculate delay before retry using exponential backoff with jitter.

        Args:
            attempt: Current attempt number (0-indexed)
            error: The exception that triggered the retry

        Returns:
            Delay in seconds
        """
        # Check for rate limit retry-after header
        if isinstance(error, RateLimitError) and error.retry_after:
            try:
                return float(error.retry_after)
            except (TypeError, ValueError):
                pass

        # Calculate exponential backoff
        delay = self.base_delay * (self.exponential_base ** attempt)

        # Apply max delay cap
        delay = min(delay, self.max_delay)

        # Apply jitter if enabled (±25%)
        if self.jitter:
            jitter_range = delay * 0.25
            delay = delay + random.uniform(-jitter_range, jitter_range)

        return max(0, delay)  # Ensure non-negative

    def should_retry(self, error: Exception, attempt: int) -> bool:
        """Determine if a request should be retried.

        Args:
            error: The exception that occurred
            attempt: Current attempt number (0-indexed)

        Returns:
            True if should retry, False otherwise
        """
        if attempt >= self.max_attempts - 1:
            return False

        # Always retry on network errors
        if isinstance(error, (NetworkError, TimeoutException, NetworkConnectionError)):
            return True

        # Retry on server errors
        if isinstance(error, ServerError):
            return True

        # Retry on rate limit
        if isinstance(error, RateLimitError):
            return True

        # Check if it's an API error with a retryable status
        if isinstance(error, APIError) and error.status_code in self.retry_on_status:
            return True

        return False


# Default retry configuration
DEFAULT_RETRY_CONFIG = RetryConfig()

# User-friendly error messages
USER_MESSAGES: Dict[str, str] = {
    'AUTHENTICATION_ERROR': 'Invalid email or password. Please try again.',
    'TOKEN_ERROR': 'Your session is invalid. Please sign in again.',
    'EMAIL_NOT_VERIFIED': 'Please verify your email address to continue.',
    'MFA_REQUIRED': 'Please complete two-factor authentication.',
    'PASSWORD_EXPIRED': 'Your password has expired. Please reset it.',
    'ACCOUNT_LOCKED': 'Your account is temporarily locked. Please try again later.',
    'SESSION_EXPIRED': 'Your session has expired. Please sign in again.',
    'AUTHORIZATION_ERROR': "You don't have permission to perform this action.",
    'INSUFFICIENT_PERMISSIONS': 'You need additional permissions for this action.',
    'VALIDATION_ERROR': 'Please check your input and try again.',
    'NOT_FOUND_ERROR': 'The requested resource was not found.',
    'CONFLICT_ERROR': 'This action conflicts with existing data.',
    'RATE_LIMIT_ERROR': 'Too many requests. Please wait a moment and try again.',
    'INTERNAL_ERROR': 'An unexpected error occurred. Please try again later.',
    'NETWORK_ERROR': 'Unable to connect. Please check your internet connection.',
}


def get_user_message(error: Union[Exception, str]) -> str:
    """Get a user-friendly error message.

    Args:
        error: Exception or error code string

    Returns:
        User-friendly message
    """
    if isinstance(error, str):
        return USER_MESSAGES.get(error, 'An unexpected error occurred.')

    if isinstance(error, JanuaError) and error.code:
        return USER_MESSAGES.get(error.code, error.message)

    if isinstance(error, Exception):
        return str(error)

    return 'An unexpected error occurred.'


class HTTPClient:
    """HTTP client for making API requests to the Janua API."""

    DEFAULT_TIMEOUT = 30.0

    def __init__(
        self,
        base_url: str,
        api_key: Optional[str] = None,
        timeout: float = DEFAULT_TIMEOUT,
        retry_config: Optional[RetryConfig] = None,
        custom_headers: Optional[Dict[str, str]] = None,
        # Legacy parameters for backwards compatibility
        max_retries: Optional[int] = None,
        retry_delay: Optional[float] = None,
    ):
        """
        Initialize the HTTP client.

        Args:
            base_url: Base URL for the API
            api_key: API key for authentication
            timeout: Request timeout in seconds
            retry_config: Retry configuration (preferred)
            custom_headers: Additional headers to include in requests
            max_retries: (Deprecated) Use retry_config instead
            retry_delay: (Deprecated) Use retry_config instead
        """
        self.base_url = base_url.rstrip('/')
        self.api_key = api_key
        self.timeout = timeout

        # Handle legacy parameters
        if retry_config:
            self.retry_config = retry_config
        elif max_retries is not None or retry_delay is not None:
            self.retry_config = RetryConfig(
                max_attempts=max_retries or DEFAULT_RETRY_CONFIG.max_attempts,
                base_delay=retry_delay or DEFAULT_RETRY_CONFIG.base_delay,
            )
        else:
            self.retry_config = DEFAULT_RETRY_CONFIG

        # Keep legacy attributes for backwards compatibility
        self.max_retries = self.retry_config.max_attempts
        self.retry_delay = self.retry_config.base_delay
        
        # Set up default headers
        self.headers = {
            'Content-Type': 'application/json',
            'Accept': 'application/json',
            'User-Agent': 'Janua-Python-SDK/1.0.0',
        }
        
        if api_key:
            self.headers['Authorization'] = f'Bearer {api_key}'
        
        if custom_headers:
            self.headers.update(custom_headers)
        
        # Create HTTP client with connection pooling
        self.client = httpx.Client(
            timeout=httpx.Timeout(timeout),
            headers=self.headers,
            follow_redirects=True,
        )
    
    def _get_url(self, endpoint: str) -> str:
        """Construct full URL for an endpoint."""
        endpoint = endpoint.lstrip('/')
        return f"{self.base_url}/{endpoint}"
    
    def _handle_response_error(self, response: Response) -> None:
        """
        Handle HTTP response errors.
        
        Args:
            response: The HTTP response object
            
        Raises:
            Various JanuaError subclasses based on status code
        """
        try:
            error_data = response.json()
            message = error_data.get('message', response.text)
            code = error_data.get('code')
            details = error_data.get('details')
        except Exception:
            message = response.text or f"HTTP {response.status_code}"
            code = None
            details = None
        
        status_code = response.status_code
        
        # Map status codes to specific exceptions
        if status_code == 401:
            raise AuthenticationError(message, code=code, details=details)
        elif status_code == 403:
            raise AuthorizationError(message, code=code, details=details)
        elif status_code == 404:
            raise NotFoundError(message, code=code, details=details)
        elif status_code == 422 or status_code == 400:
            raise ValidationError(message, code=code, details=details)
        elif status_code == 429:
            # Extract retry-after header if available
            retry_after = response.headers.get('Retry-After')
            details = details or {}
            if retry_after:
                details['retry_after'] = retry_after
            raise RateLimitError(message, code=code, details=details)
        elif 500 <= status_code < 600:
            raise ServerError(message, code=code, details=details)
        else:
            raise APIError(message, status_code=status_code, code=code, details=details)
    
    def _should_retry(self, error: Exception, attempt: int) -> bool:
        """
        Determine if a request should be retried.

        Args:
            error: The exception that occurred
            attempt: Current attempt number (0-indexed)

        Returns:
            True if should retry, False otherwise
        """
        return self.retry_config.should_retry(error, attempt)

    def _calculate_retry_delay(self, attempt: int, error: Optional[Exception] = None) -> float:
        """
        Calculate delay before retry using exponential backoff.

        Args:
            attempt: Current attempt number (0-indexed)
            error: The exception that occurred

        Returns:
            Delay in seconds
        """
        return self.retry_config.calculate_delay(attempt, error)
    
    def request(
        self,
        method: str,
        endpoint: str,
        json: Optional[Dict[str, Any]] = None,
        params: Optional[Dict[str, Any]] = None,
        headers: Optional[Dict[str, str]] = None,
        **kwargs
    ) -> Response:
        """
        Make an HTTP request with retry logic.
        
        Args:
            method: HTTP method (GET, POST, PUT, DELETE, etc.)
            endpoint: API endpoint path
            json: JSON body for the request
            params: Query parameters
            headers: Additional headers for this request
            **kwargs: Additional arguments to pass to httpx
            
        Returns:
            HTTP response object
            
        Raises:
            Various JanuaError subclasses based on error type
        """
        url = self._get_url(endpoint)
        
        # Merge headers
        request_headers = self.headers.copy()
        if headers:
            request_headers.update(headers)
        
        attempt = 0
        last_error = None

        while attempt < self.retry_config.max_attempts:
            try:
                response = self.client.request(
                    method=method,
                    url=url,
                    json=json,
                    params=params,
                    headers=request_headers,
                    **kwargs
                )
                
                # Check for HTTP errors
                if response.status_code >= 400:
                    self._handle_response_error(response)
                
                return response
                
            except (NetworkError, TimeoutException) as e:
                last_error = NetworkConnectionError(
                    f"Network error during {method} {url}: {str(e)}"
                )
                
                if not self._should_retry(last_error, attempt):
                    raise last_error
                
            except (AuthenticationError, AuthorizationError, NotFoundError, ValidationError):
                # Don't retry client errors
                raise
                
            except (RateLimitError, ServerError, APIError) as e:
                last_error = e
                
                if not self._should_retry(e, attempt):
                    raise
                
            except Exception as e:
                # Unexpected error
                raise JanuaError(f"Unexpected error during {method} {url}: {str(e)}")
            
            # Calculate retry delay and wait
            if last_error and attempt < self.retry_config.max_attempts - 1:
                delay = self._calculate_retry_delay(attempt, last_error)

                # Call on_retry callback if configured
                if self.retry_config.on_retry:
                    self.retry_config.on_retry(attempt, last_error, delay)

                time.sleep(delay)

            attempt += 1

        # If we get here, we've exhausted retries
        if last_error:
            raise last_error
        else:
            raise JanuaError(f"Failed to complete request after {self.retry_config.max_attempts} retries")
    
    def get(
        self,
        endpoint: str,
        params: Optional[Dict[str, Any]] = None,
        **kwargs
    ) -> Response:
        """Make a GET request."""
        return self.request('GET', endpoint, params=params, **kwargs)
    
    def post(
        self,
        endpoint: str,
        json: Optional[Dict[str, Any]] = None,
        **kwargs
    ) -> Response:
        """Make a POST request."""
        return self.request('POST', endpoint, json=json, **kwargs)
    
    def put(
        self,
        endpoint: str,
        json: Optional[Dict[str, Any]] = None,
        **kwargs
    ) -> Response:
        """Make a PUT request."""
        return self.request('PUT', endpoint, json=json, **kwargs)
    
    def patch(
        self,
        endpoint: str,
        json: Optional[Dict[str, Any]] = None,
        **kwargs
    ) -> Response:
        """Make a PATCH request."""
        return self.request('PATCH', endpoint, json=json, **kwargs)
    
    def delete(
        self,
        endpoint: str,
        **kwargs
    ) -> Response:
        """Make a DELETE request."""
        return self.request('DELETE', endpoint, **kwargs)
    
    def close(self) -> None:
        """Close the HTTP client and release resources."""
        self.client.close()
    
    def __enter__(self):
        """Context manager entry."""
        return self
    
    def __exit__(self, exc_type, exc_val, exc_tb):
        """Context manager exit."""
        self.close()

File: packages/flutter-sdk/lib/src/http_client.dart
/// HTTP client for the Janua Flutter SDK.
import 'dart:async';
import 'dart:convert';
import 'dart:io';
import 'dart:math';
import 'package:http/http.dart' as http;
import 'config.dart';
import 'exceptions/janua_exception.dart';

/// Configuration for retry behavior.
class RetryConfig {
  /// Maximum number of retry attempts (including initial request).
  final int maxAttempts;

  /// Initial delay between retries in milliseconds.
  final int baseDelayMs;

  /// Maximum delay between retries in milliseconds.
  final int maxDelayMs;

  /// Base for exponential backoff (default: 2.0).
  final double exponentialBase;

  /// Whether to add random jitter to delay (prevents thundering herd).
  final bool jitter;

  /// HTTP status codes that should trigger a retry.
  final Set<int> retryOnStatusCodes;

  /// Callback called before each retry.
  final void Function(int attempt, Exception error, Duration delay)? onRetry;

  const RetryConfig({
    this.maxAttempts = 3,
    this.baseDelayMs = 1000,
    this.maxDelayMs = 60000,
    this.exponentialBase = 2.0,
    this.jitter = true,
    this.retryOnStatusCodes = const {429, 500, 502, 503, 504},
    this.onRetry,
  });

  /// Calculate delay before retry using exponential backoff with jitter.
  Duration calculateDelay(int attempt, {Duration? retryAfter}) {
    // Use retry-after header if provided (for rate limits)
    if (retryAfter != null) {
      return retryAfter;
    }

    // Calculate exponential backoff
    double delayMs = baseDelayMs * pow(exponentialBase, attempt).toDouble();

    // Apply max delay cap
    delayMs = min(delayMs, maxDelayMs.toDouble());

    // Apply jitter if enabled (±25%)
    if (jitter) {
      final random = Random();
      final jitterRange = delayMs * 0.25;
      delayMs = delayMs + (random.nextDouble() * 2 - 1) * jitterRange;
    }

    return Duration(milliseconds: max(0, delayMs.toInt()));
  }

  /// Determine if an error should trigger a retry.
  bool shouldRetry(Exception error, int attempt) {
    if (attempt >= maxAttempts - 1) {
      return false;
    }

    // Always retry on network errors
    if (error is NetworkException || error is SocketException) {
      return true;
    }

    // Retry on rate limit
    if (error is RateLimitException) {
      return true;
    }

    // Check if it's a server error with retryable status
    if (error is JanuaException && error.statusCode != null) {
      return retryOnStatusCodes.contains(error.statusCode);
    }

    return false;
  }

  /// Creates a copy with modified values.
  RetryConfig copyWith({
    int? maxAttempts,
    int? baseDelayMs,
    int? maxDelayMs,
    double? exponentialBase,
    bool? jitter,
    Set<int>? retryOnStatusCodes,
    void Function(int, Exception, Duration)? onRetry,
  }) {
    return RetryConfig(
      maxAttempts: maxAttempts ?? this.maxAttempts,
      baseDelayMs: baseDelayMs ?? this.baseDelayMs,
      maxDelayMs: maxDelayMs ?? this.maxDelayMs,
      exponentialBase: exponentialBase ?? this.exponentialBase,
      jitter: jitter ?? this.jitter,
      retryOnStatusCodes: retryOnStatusCodes ?? this.retryOnStatusCodes,
      onRetry: onRetry ?? this.onRetry,
    );
  }
}

/// Default retry configuration.
const defaultRetryConfig = RetryConfig();

/// User-friendly error messages for common error codes.
const Map<String, String> userMessages = {
  'AUTHENTICATION_ERROR': 'Invalid email or password. Please try again.',
  'TOKEN_ERROR': 'Your session is invalid. Please sign in again.',
  'EMAIL_NOT_VERIFIED': 'Please verify your email address to continue.',
  'MFA_REQUIRED': 'Please complete two-factor authentication.',
  'PASSWORD_EXPIRED': 'Your password has expired. Please reset it.',
  'ACCOUNT_LOCKED': 'Your account is temporarily locked. Please try again later.',
  'SESSION_EXPIRED': 'Your session has expired. Please sign in again.',
  'AUTHORIZATION_ERROR': "You don't have permission to perform this action.",
  'INSUFFICIENT_PERMISSIONS': 'You need additional permissions for this action.',
  'VALIDATION_ERROR': 'Please check your input and try again.',
  'NOT_FOUND': 'The requested resource was not found.',
  'CONFLICT_ERROR': 'This action conflicts with existing data.',
  'RATE_LIMIT_EXCEEDED': 'Too many requests. Please wait a moment and try again.',
  'INTERNAL_ERROR': 'An unexpected error occurred. Please try again later.',
  'NETWORK_ERROR': 'Unable to connect. Please check your internet connection.',
};

/// Get a user-friendly error message for an exception.
String getUserMessage(JanuaException error) {
  return userMessages[error.code] ?? error.message;
}

/// HTTP client wrapper for making API requests.
class JanuaHttpClient {
  final JanuaConfig config;
  final http.Client _client;
  final RetryConfig retryConfig;

  String? _accessToken;

  JanuaHttpClient({
    required this.config,
    http.Client? client,
    RetryConfig? retryConfig,
  })  : _client = client ?? http.Client(),
        retryConfig = retryConfig ?? defaultRetryConfig;

  /// Sets the access token for authenticated requests.
  void setAccessToken(String? token) {
    _accessToken = token;
  }

  /// Clears the access token.
  void clearAccessToken() {
    _accessToken = null;
  }

  /// Makes a GET request.
  Future<Map<String, dynamic>> get(
    String path, {
    Map<String, String>? queryParameters,
    bool authenticated = true,
  }) async {
    return _request('GET', path,
        queryParameters: queryParameters, authenticated: authenticated);
  }

  /// Makes a POST request.
  Future<Map<String, dynamic>> post(
    String path, {
    Map<String, dynamic>? body,
    bool authenticated = true,
  }) async {
    return _request('POST', path, body: body, authenticated: authenticated);
  }

  /// Makes a PUT request.
  Future<Map<String, dynamic>> put(
    String path, {
    Map<String, dynamic>? body,
    bool authenticated = true,
  }) async {
    return _request('PUT', path, body: body, authenticated: authenticated);
  }

  /// Makes a PATCH request.
  Future<Map<String, dynamic>> patch(
    String path, {
    Map<String, dynamic>? body,
    bool authenticated = true,
  }) async {
    return _request('PATCH', path, body: body, authenticated: authenticated);
  }

  /// Makes a DELETE request.
  Future<Map<String, dynamic>> delete(
    String path, {
    Map<String, dynamic>? body,
    bool authenticated = true,
  }) async {
    return _request('DELETE', path, body: body, authenticated: authenticated);
  }

  Future<Map<String, dynamic>> _request(
    String method,
    String path, {
    Map<String, dynamic>? body,
    Map<String, String>? queryParameters,
    bool authenticated = true,
  }) async {
    final uri = _buildUri(path, queryParameters);
    final headers = _buildHeaders(authenticated: authenticated);

    if (config.debug) {
      _logRequest(method, uri, headers, body);
    }

    int attempt = 0;
    Exception? lastError;
    Duration? retryAfterDuration;

    while (attempt < retryConfig.maxAttempts) {
      try {
        http.Response response;
        final timeout = Duration(seconds: config.timeoutSeconds);

        switch (method) {
          case 'GET':
            response = await _client.get(uri, headers: headers).timeout(timeout);
            break;
          case 'POST':
            response = await _client
                .post(uri, headers: headers, body: body != null ? jsonEncode(body) : null)
                .timeout(timeout);
            break;
          case 'PUT':
            response = await _client
                .put(uri, headers: headers, body: body != null ? jsonEncode(body) : null)
                .timeout(timeout);
            break;
          case 'PATCH':
            response = await _client
                .patch(uri, headers: headers, body: body != null ? jsonEncode(body) : null)
                .timeout(timeout);
            break;
          case 'DELETE':
            response = await _client.delete(uri, headers: headers).timeout(timeout);
            break;
          default:
            throw ArgumentError('Unsupported HTTP method: $method');
        }

        if (config.debug) {
          _logResponse(response);
        }

        // Parse Retry-After header if present (for rate limiting)
        retryAfterDuration = _parseRetryAfter(response.headers);

        return _handleResponse(response);
      } on SocketException catch (e) {
        lastError = NetworkException(message: 'Network error: ${e.message}');
        if (!retryConfig.shouldRetry(lastError, attempt)) {
          throw lastError;
        }
      } on TimeoutException catch (_) {
        lastError = NetworkException(message: 'Request timeout');
        if (!retryConfig.shouldRetry(lastError, attempt)) {
          throw lastError;
        }
      } on http.ClientException catch (e) {
        lastError = NetworkException(message: 'HTTP client error: ${e.message}');
        if (!retryConfig.shouldRetry(lastError, attempt)) {
          throw lastError;
        }
      } on FormatException catch (e) {
        // Don't retry format exceptions
        throw JanuaException(
          message: 'Invalid response format: ${e.message}',
          code: 'INVALID_RESPONSE',
        );
      } on AuthenticationException {
        // Don't retry auth errors
        rethrow;
      } on AuthorizationException {
        // Don't retry authorization errors
        rethrow;
      } on ValidationException {
        // Don't retry validation errors
        rethrow;
      } on NotFoundException {
        // Don't retry not found errors
        rethrow;
      } on RateLimitException catch (e) {
        lastError = e;
        if (!retryConfig.shouldRetry(e, attempt)) {
          rethrow;
        }
        // Use retry-after from exception if available
        if (e.retryAfter != null) {
          retryAfterDuration = e.retryAfter!.difference(DateTime.now());
          if (retryAfterDuration.isNegative) {
            retryAfterDuration = null;
          }
        }
      } on JanuaException catch (e) {
        lastError = e;
        if (!retryConfig.shouldRetry(e, attempt)) {
          rethrow;
        }
      }

      // Calculate delay and wait before retry
      if (attempt < retryConfig.maxAttempts - 1) {
        final delay = retryConfig.calculateDelay(attempt, retryAfter: retryAfterDuration);

        if (config.debug) {
          // ignore: avoid_print
          print('[Janua] Retry attempt ${attempt + 1}/${retryConfig.maxAttempts} after ${delay.inMilliseconds}ms');
        }

        // Call on_retry callback if configured
        if (retryConfig.onRetry != null && lastError != null) {
          retryConfig.onRetry!(attempt, lastError, delay);
        }

        await Future.delayed(delay);
      }

      attempt++;
      retryAfterDuration = null; // Reset for next attempt
    }

    // If we get here, we've exhausted retries
    if (lastError != null) {
      throw lastError;
    }
    throw JanuaException(
      message: 'Failed to complete request after ${retryConfig.maxAttempts} attempts',
      code: 'MAX_RETRIES_EXCEEDED',
    );
  }

  /// Parses the Retry-After header from response headers.
  Duration? _parseRetryAfter(Map<String, String> headers) {
    final retryAfter = headers['retry-after'] ?? headers['Retry-After'];
    if (retryAfter == null) return null;

    // Try to parse as seconds (integer)
    final seconds = int.tryParse(retryAfter);
    if (seconds != null) {
      return Duration(seconds: seconds);
    }

    // Try to parse as HTTP-date
    try {
      final date = HttpDate.parse(retryAfter);
      final duration = date.difference(DateTime.now());
      return duration.isNegative ? null : duration;
    } catch (_) {
      return null;
    }
  }

  Uri _buildUri(String path, Map<String, String>? queryParameters) {
    final baseUri = Uri.parse(config.baseUrl);
    final fullPath = path.startsWith('/') ? '/api/v1$path' : '/api/v1/$path';

    return Uri(
      scheme: baseUri.scheme,
      host: baseUri.host,
      port: baseUri.port,
      path: fullPath,
      queryParameters: queryParameters?.isNotEmpty == true ? queryParameters : null,
    );
  }

  Map<String, String> _buildHeaders({bool authenticated = true}) {
    final headers = <String, String>{
      'Content-Type': 'application/json',
      'Accept': 'application/json',
      'X-Tenant-ID': config.tenantId,
      'X-Client-ID': config.clientId,
      'X-SDK': 'flutter',
      'X-SDK-Version': '1.0.0',
    };

    if (authenticated && _accessToken != null) {
      headers['Authorization'] = 'Bearer $_accessToken';
    }

    if (config.customHeaders != null) {
      headers.addAll(config.customHeaders!);
    }

    return headers;
  }

  Map<String, dynamic> _handleResponse(http.Response response) {
    final statusCode = response.statusCode;
    Map<String, dynamic> data;

    try {
      if (response.body.isEmpty) {
        data = {'success': true};
      } else {
        data = jsonDecode(response.body) as Map<String, dynamic>;
      }
    } catch (_) {
      data = {'message': response.body};
    }

    if (statusCode >= 200 && statusCode < 300) {
      return data;
    }

    // Handle errors
    final message = data['message'] as String? ?? 'Unknown error';
    final code = data['code'] as String? ?? 'UNKNOWN_ERROR';
    final details = data['details'] as Map<String, dynamic>?;

    switch (statusCode) {
      case 400:
        throw ValidationException(
          message: message,
          code: code,
          details: details,
          statusCode: statusCode,
        );
      case 401:
        throw AuthenticationException(
          message: message,
          code: code,
          details: details,
          statusCode: statusCode,
        );
      case 403:
        throw AuthorizationException(
          message: message,
          code: code,
          details: details,
          statusCode: statusCode,
        );
      case 404:
        throw NotFoundException(
          message: message,
          code: code,
          details: details,
          statusCode: statusCode,
        );
      case 429:
        throw RateLimitException(
          message: message,
          code: code,
          details: details,
          statusCode: statusCode,
        );
      default:
        throw JanuaException(
          message: message,
          code: code,
          details: details,
          statusCode: statusCode,
        );
    }
  }

  void _logRequest(
    String method,
    Uri uri,
    Map<String, String> headers,
    Map<String, dynamic>? body,
  ) {
    // ignore: avoid_print
    print('[Janua] $method $uri');
    if (body != null) {
      // ignore: avoid_print
      print('[Janua] Body: ${jsonEncode(body)}');
    }
  }

  void _logResponse(http.Response response) {
    // ignore: avoid_print
    print('[Janua] Response ${response.statusCode}: ${response.body}');
  }

  /// Closes the HTTP client.
  void close() {
    _client.close();
  }
}
